subprojects/core-api/src/main/java/org/gradle/internal/deprecation/DeprecatableConfiguration.java

```diff
@@ -46,20 +46,24 @@ public interface DeprecatableConfiguration extends Configuration {
     List<String> getResolutionAlternatives();
 
     /**
-     * Allows plugins to deprecate a configuration that will be removed in the next major Gradle version.
+     * Allows plugins to deprecate the consumability property (canBeConsumed() == true) of a configuration that will be changed in the next major Gradle version.
      *
-     * @param alternativesForDeclaring alternative configurations that can be used to declare dependencies
+     * @param deprecation deprecation message builder to use for nagging upon consumption of this configuration
      * @return this configuration
      */
-    DeprecatableConfiguration deprecateForDeclaration(String... alternativesForDeclaring);
+    DeprecatableConfiguration deprecateForConsumption(Function<DeprecationMessageBuilder.DeprecateConfiguration, DeprecationMessageBuilder.WithDocumentation> deprecation);
```

Comment:
I think we should leave these in place for now.  Though they are an internal API, they are a very useful toolkit for plugin authors, and us to track improper usage and migrate.  Which I know sounds similar to the point I conceded on undeprecation, but I think this is on much firmer ground.

Reply:
If these are methods we expect plugin authors to be calling, we should at the very least make this a public API. We run into enough issues with plugins calling our internal APIs and having code in this grey-area makes things very confusing from a deprecation cycle perspective. 

Though, your argument is confusing to me since this entire PR is setting up a framework to managing roles and deprecations of configurations -- which occurs at configuration construction time. Its not like removing these methods disallows deprecating configurations. That functionality has just been replaced by the new API you're writing here.

Reply:
I'm thinking authors of plugin A might want to mark configurations created by plugin B (which A applies) as deprecated, so that they can correct A's own improper usage of these configurations which they do not create.  

It's grey-area code, but again, it's strictly time limited and probably already done.  If we move these to public API then we'll have to go through a deprecation cycle to remove them, I don't want to get held up with that - there already enough constraints on how fast we can go in the projected phases in the spec.